akka-stream-tests/src/test/scala/akka/stream/impl/fusing/InterpreterSpec.scala

```diff
@@ -569,39 +569,67 @@ class InterpreterSpec extends AkkaSpec with GraphInterpreterSpecKit {
 
   }
 
-  private[akka] case class Doubler[T]() extends PushPullStage[T, T] {
-    var oneMore: Boolean = false
-    var lastElem: T = _
-
-    override def onPush(elem: T, ctx: Context[T]): SyncDirective = {
-      lastElem = elem
-      oneMore = true
-      ctx.push(elem)
-    }
+  private[akka] final case class Doubler[T]() extends GraphStage[FlowShape[T, T]] {
```

Comment:
Current branch is checked out from an existing yet not merged pr #20307. I'm planning to rebase this branch when the other pr get merged (I'll and a `wip` prefix to this pr's title). And your suggestion sounds great to me, I'll separate each stage into different PR for the later ones (see #20288).  Do these sound ok?